ballista/scheduler/src/flight_proxy_service.rs

```diff
@@ -0,0 +1,157 @@
+        &self,
+        _request: Request<Streaming<HandshakeRequest>>,
+    ) -> Result<Response<Self::HandshakeStream>, Status> {
+        Err(Status::unimplemented("handshake"))
+    }
+
+    async fn list_flights(
+        &self,
+        _request: Request<Criteria>,
+    ) -> Result<Response<Self::ListFlightsStream>, Status> {
+        Err(Status::unimplemented("list_flights"))
+    }
+
+    async fn get_flight_info(
+        &self,
+        _request: Request<FlightDescriptor>,
+    ) -> Result<Response<FlightInfo>, Status> {
+        Err(Status::unimplemented("get_flight_info"))
+    }
+
+    async fn poll_flight_info(
+        &self,
+        _request: Request<FlightDescriptor>,
+    ) -> Result<Response<PollInfo>, Status> {
+        Err(Status::unimplemented("poll_flight_info"))
+    }
+
+    async fn get_schema(
+        &self,
+        _request: Request<FlightDescriptor>,
+    ) -> Result<Response<SchemaResult>, Status> {
+        Err(Status::unimplemented("get_schema"))
+    }
+
+    async fn do_get(
+        &self,
+        request: Request<Ticket>,
+    ) -> Result<Response<Self::DoGetStream>, Status> {
+        let ticket = request.into_inner();
+
+        let action =
+            decode_protobuf(&ticket.ticket).map_err(|e| from_ballista_err(&e))?;
+
+        match &action {
+            BallistaAction::FetchPartition {
+                host, port, job_id, ..
+            } => {
+                debug!("Fetching results for job id: {job_id} from {host}:{port}");
+                let mut client = get_flight_client(host, port)
+                    .map_err(|e| from_ballista_err(&e))
+                    .await?;
+                client
+                    .do_get(Request::new(ticket))
+                    .await
+                    .map(|r| Response::new(Box::pin(r.into_inner()) as Self::DoGetStream))
+            }
+        }
+    }
+
+    async fn do_put(
+        &self,
+        _request: Request<Streaming<FlightData>>,
+    ) -> Result<Response<Self::DoPutStream>, Status> {
+        Err(Status::unimplemented("do_put"))
+    }
+
+    async fn do_exchange(
+        &self,
+        _request: Request<Streaming<FlightData>>,
+    ) -> Result<Response<Self::DoExchangeStream>, Status> {
+        Err(Status::unimplemented("do_exchange"))
+    }
+
+    async fn do_action(
+        &self,
+        _request: Request<Action>,
+    ) -> Result<Response<Self::DoActionStream>, Status> {
+        Err(Status::unimplemented("do_action"))
+    }
+
+    async fn list_actions(
+        &self,
+        _request: Request<Empty>,
+    ) -> Result<Response<Self::ListActionsStream>, Status> {
+        Err(Status::unimplemented("list_actions"))
+    }
+}
+
+fn from_ballista_err(e: &ballista_core::error::BallistaError) -> Status {
+    Status::internal(format!("Ballista Error: {e:?}"))
+}
+
+async fn get_flight_client(
+    host: &String,
+    port: &u16,
+) -> Result<FlightServiceClient<tonic::transport::channel::Channel>, BallistaError> {
```

Comment:
Consider connection pooling. Currently a new client is created for each request.